### DISCUSSION TITLE
Implementing getters and setters in the language server

### DISCUSSION THREAD

**Discussion by dantleech** [phpactor/phpactor#1244] [Ideas]
How to implement getters / setters

**Comment by dantleech** [comment]
@weeman1337 the code in general is [here](https://github.com/phpactor/language-server-phpactor-extensions/tree/master/lib/LanguageServerCodeTransform/CodeAction).

There is more than one approach here, one would be to:

- Implement a [command](https://github.com/phpactor/language-server-phpactor-extensions/blob/master/lib/LanguageServerCodeTransform/LspCommand/ImportNameCommand.php) which will generate the accessors.

Then:

- Create a [code action](https://github.com/phpactor/language-server-phpactor-extensions/blob/master/lib/LanguageServerCodeTransform/CodeAction/ImportNameProvider.php) which detects if the command would be applicable, and if so, return a code action referencing the command

The other would be to implement a [Transformer](https://github.com/phpactor/code-transform/blob/d6ebbc1472682dbe1e0b5b6e4bc8c6b9d46caccd/lib/Adapter/TolerantParser/ClassToFile/Transformer/ClassNameFixerTransformer.php) in `code-transform`, which could then plug into the existing `TransfotrmerCodeActionProvider`. One benefit with this approach is that it will apply equally to the legacy system.

**Comment by weeman1337** [comment]
Thanks for the hints.

In my opinion the final picture should be
- In property context: generate accessor / mutator / both (single one for that property)
- In class context: generate accessors / mutators / both (all for class)

My questions?
- Makes that above sense?
- How shout it be named?
  - accessor(s) / mutator(s)
  - or getter(s) / setter(s)

**Reply by dantleech** [reply]
sorry missed the notification for this. yes - that makes sense - the problem at the moment is that last time I checked LSP code actions do not provide the offset/position at which they were invoked, so we do/did not have the context.

I notice that in `tsserver` however you can invoke code actions at specific points and get offsets, so maybe this has chagned or I missed it the first time round.

**Reply by dantleech** [reply]
My original suggestion above relates to a global "transform" operation on the code, but having a contextual code action would be better.

**Comment by BladeMF** [comment]
Should we discuss how to migrate everything [here](https://github.com/phpactor/code-transform/tree/master/lib/Adapter/WorseReflection) from returning `SourceCode` to returning `TextEdits` so we can invoke it in LSP?

**Reply by dantleech** [reply]
theoreticaly it's quite easy - most of them convert Phpactor TextEdits to a `SourceCode`. So it's a case of returning the TextEdits directlry and updating the call sites to do `$textEdits->apply($source)`

**Reply by BladeMF** [reply]
Yes, but I think we should start with this one. Deciding to update all call sites is an easy call to make. If you want, let's get this out to another discussion for further reference.

**Reply by dantleech** [reply]
it's fine, can discuss here.

**Reply by BladeMF** [reply]
I created a new one. It will be more easily findable for newcomers I think.
